---
title: "Assignment Week 3"
author: "Leonard C. Steg"
date: "3/16/2022"
output: html_document
---

# Load libraries, set wd and load AnnotationHub
```{r}
suppressPackageStartupMessages({
  library(AnnotationHub)
  library(Rsubread)
  library(rtracklayer)
  library(Biostrings)
  library(Rfastp)
  library(epiwraps)
})

setwd("/mnt/groupMansuy/leo/epibioinfo/week03/")
ah <- AnnotationHub()
```

# Downloading  raw reads
```{r, eval=FALSE}

if (!dir.exists("raw")){
dir.create("raw")
} 

download.file("https://www.encodeproject.org/files/ENCFF001LJN/@@download/ENCFF001LJN.fastq.gz", dest="raw/mESCs.p300.fastq.gz")
```

# QC and trimming

## Using Rfastp

```{r, message=FALSE}
dir.create("rfastp.trimmed")
if (!dir.exists("rfastp.trimmed")){
dir.create("rfastp.trimmed")
} 
qc <- Rfastp::rfastp("raw/mESCs.p300.fastq.gz", outputFastq="rfastp.trimmed/mESC.p300", 
                     maxNfilter=0, thread=4, overrepresentationAnalysis=TRUE)
```
### Plot Curve plots before and after QC
```{r}
Rfastp::curvePlot(qc, curve="content_curves")
```


# Alignment
```{r, eval=F, echo=T}
# Load the mouse genome sequence
genome <- ah[["AH49775"]]

# Export the genome sequence in fasta format
if (!dir.exists("mm_genome")){
dir.create("mm_genome")
} 
export(import.2bit(genome), "mm_genome/genome.fasta.gz", compress=TRUE)
```

### Building the index using Bowtie2
```{bash engine.opts='-l', eval=F, echo=T}
bowtie2-build --threads 4 /mnt/groupMansuy/week03/mm_genome/genome.fasta.gz /mnt/groupMansuy/week03/mm_genome/bowtie2
```

### Alignment using samtools
```{bash engine.opts='-l', eval=F, echo=T}
mkdir -p aligned
(bowtie2 -p 4 -x mm_genome/bowtie2 -U rfastp.trimmed/mESC.p300_R1.fastq.gz) 2> aligned/mESC.p300.bowtie2 |\
  samtools view -bS - | samtools sort -@4 -m 2G - > aligned/mESC.p300.bam
samtools index aligned/mESC.p300.bam
```


# Peak calling

### Using MACS2

```{bash engine.opts='-l', eval=F, echo=T}
mkdir -p peaks
macs2 callpeak --outdir peaks -n mESC.p300 --gsize dm -t aligned/mESC.p300.bam
```

### Importing peaks into R

```{r}
peaks <- rtracklayer::import("peaks/mESC.p300_peaks.narrowPeak", format = "narrowPeak")
```

### Assignment question: How many peaks do we find?
```{r}
length(peaks)
```


# Generating coverage tracks

```{r, eval=F, echo=T}
dir.create("tracks")
if (!dir.exists("tracks")){
dir.create("tracks")
} 
bam2bw("aligned/mESC.p300.bam", "tracks/mESC.p300.bw", binWidth=10, extend=50, scaling=TRUE)
```
# Load Mouse EnsDb to have gene annoation

```{r}
ensdb <- ah[["AH89211"]] # Mouse EnsDb
```


# Looking at the p300 peak(s) around the Sox9 gene

```{r}
plotSignalTracks(list(signal="tracks/mESC.p300.bw", peaks=peaks), region="Sox9", ensdb=ensdb, colors = "#017788", transcripts = "collapsed")
```



### Add gene annotaiton to peak data and describe annotation of identified genes
```{r}
peakAnno <- annotateRegions(peaks, ensdb)
table(peakAnno$class)
```

```{r}
sessionInfo()
```

